### PR TITLE
Integrate LibWallet into the `tari_console_wallet` application

### PR DESCRIPTION
**Pull request by philipr-za**
## Description
This PR integrates the tari_wallet backend into the tari_console_wallet application. The application needed to reused a lot of the utility functions from `tari_base_node` so it moves those utility functions into a new crate called `tari_app_utilities`. 

The wallet will use the same config location as `tari_base_node` and the same identity and data files. It will create a separate log config and log to a separate folder within ~/.tari.

This PR is still in progress and the following is what I think still needs to be done to flesh out all the functionality present in the current UI:

- Add way to specific wallet base node in config (Currently just using the first Seed Node)
- Add way to edit base node pub key and address in console UI
- Read Balance from wallet
- Read Contacts from wallet
- Update above on relevant wallet events by subscribing to the wallet Event Streams.
- Read Chain tip from Base Node in wallet (this is a whole thing) and plumb into UI with updates via wallet events
- Add calling Send transaction in wallet
- Add contacts from UI
- Read current connections from Wallet for Network Tab
- Optimize the cacheing and display of the Complete Transaction list which lags when it is very long. Currently it caches the full completed transactions but only needs a few fields which should help a lot.
- Determine if running this app with the old BN will create issues with a single database file in terms of crashes and corruption (Perhaps the WAL journal mode is good enough to allow this operation in Sqlite?)

After this we still need to integrate gRPC and add a Console page.

## How Has This Been Tested?
Tests provided for features added to support application.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_app_grpc/src/conversions/aggregate_body.rs
Comment (sdbondi) [inline applications/tari_app_grpc/src/conversions/aggregate_body.rs]: Should rather implement `impl Into<TransactionInput> for grpc::TransactionInput` and use `.map(Into::into)` because this snippet of code occurs in a few places.

File: applications/tari_app_grpc/src/conversions/new_block_template.rs
Comment (sdbondi) [inline applications/tari_app_grpc/src/conversions/new_block_template.rs]: Reuse `impl Into<grpc::TransactionInput>`, same with outputs and kernels

File: applications/tari_app_grpc/src/conversions/aggregate_body.rs
Comment (sdbondi) [reply applications/tari_app_grpc/src/conversions/aggregate_body.rs]: Same with `grpc::TransactionOutput` and `grpc::TransactionKernel`

File: applications/tari_app_utilities/src/identity_management.rs
Comment (sdbondi) [inline applications/tari_app_utilities/src/identity_management.rs]: nit: While we're moving utility functions around I'd recommend making them generic on `AsRef<Path>` so that you can write e.g. `setup_node_identity("/tmp/file.tmp", ...)`. Same for `load_identity` - just for the sake of ease of use and consistency

```suggestion
pub fn setup_node_identity<P: AsRef<Path>>(
    identity_file: P,
```

File: applications/tari_app_utilities/src/identity_management.rs
Comment (sdbondi) [inline applications/tari_app_utilities/src/identity_management.rs]: nit:
```suggestion
pub fn load_identity<P: AsRef<Path>>(path: P) -> Result<NodeIdentity, String> {
    if !path.as_ref().exists() {
```

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs
Comment (sdbondi) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:41]: You'll be able to avoid the lock entirely by cloning the `transaction_service` each time you need to use it. Essentially you're cloning an `mpsc::UnboundedSender` which is very cheap. 

```suggestion
    pub fn new(
        wallet: Arc<
                Wallet<
                    WalletSqliteDatabase,
                    TransactionServiceSqliteDatabase,
                    OutputManagerSqliteDatabase,
                    ContactsServiceSqliteDatabase,
                >,
        >,
    ) -> Self
    {
        Self { wallet }
    }
```

Depending on how `TransactionService` is implemented, you should also be able to get some concurrency gains.

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs
Comment (sdbondi) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:57]:
```suggestion
       let mut transaction_sevice = self
            .wallet
            .transaction_service.clone();
         let response = transaction_service
            .generate_coinbase_transaction(request.reward.into(), request.fee.into(), request.height)
            .await;
```

File: applications/tari_console_wallet/src/ui/app.rs
Comment (sdbondi) [inline applications/tari_console_wallet/src/ui/app.rs:186]: Same here - rather than using a RwLock - it's cheaper and more concurrent to clone the handle and use it mutably

File: applications/tari_console_wallet/src/ui/stateful_list.rs
Comment (sdbondi) [inline applications/tari_console_wallet/src/ui/stateful_list.rs]: nit: It feels more correct to return a reference from this function i.e. `Option<&T>` and leave the cloning to the consumer if necessary - looking at usages of this function, I see `selected_item().as_ref()` which is wasteful. Granted most other places seem to need a `selected_item().cloned()`, but I prefer the explicitness anyway.

Comment (sdbondi) [reply applications/tari_console_wallet/src/ui/stateful_list.rs]:
```rust
self.state.selected().and_then(|i| self.items.get(i))
```

File: applications/tari_console_wallet/src/ui/ui_contact.rs
Comment (sdbondi) [inline applications/tari_console_wallet/src/ui/ui_contact.rs]: nit: Clippy would probably complain - `c.public_key.to_string()` 🤷‍♂️

**Review by sdbondi — COMMENTED** [review]
Looks awesome! I'm pretty sure that you can do away with the `RwLock` if you simply clone the handle and use it mutably as needed.

I'm going to run this branch and do some UI testing.

File: applications/tari_app_grpc/src/conversions/aggregate_body.rs
Comment (sdbondi) [reply applications/tari_app_grpc/src/conversions/aggregate_body.rs]: You may run into issues with this (conflicting implementations) - in which case, I say leave it as is

**Comment by sdbondi** [comment]
Looks like we need to get rid of some warnings before this can be considered to go in. Some of the warnings are more serious e.g. unused results and futures but generally all warnings should be fixed before being merged in.

File: applications/tari_app_grpc/src/conversions/aggregate_body.rs
Comment (stringhandler) [reply applications/tari_app_grpc/src/conversions/aggregate_body.rs]: This code was actually just moved from another file, not changed in this implementation. So maybe something to look at at a later stage

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs
Comment (stringhandler) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:41]: Perhaps worth looking into this at a later stage. I don't disagree that it may be more optimal, but there is still quite a lot to implement, and at this stage it's a fairly heavy lock. Conceptually I was thinking: The GRPC server and TUI will be using the wallet. Some parts of the GRPC are only going to need read access to it, some need write. In the TUI, the display cycle only reads from the wallet, and when you press a button, or invoke an action you need write access to it. You may be able to go into finer grain locking, or use a non-locking strategy, but as I said, not all of the functions are implemented yet, so you may find that you constantly changing these locks.

Comment (philipr-za) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:41]: As discussed this architectural comment will be addressed in subsequent PRs

**Comment by philipr-za** [comment]
Ok this is ready for a review. Cleaned up all the warnings and addressed most the comment. The comments that were not directly addressed will be dealt with in subsequent PRs

**Review by stringhandler — APPROVED** [review]

**Review by sdbondi — APPROVED** [review]

**Pull request merged**
